### ISSUE TITLE
Panzer: Kokkos thread blocked gather/scatter

### ISSUE THREAD

**Issue by bathmatt** [trilinos/Trilinos#2497]
Right now the gather and scatters are only threaded for the standard DOF, and not the blocked one, EMPIRE uses the blocked one and this causes a big slowdown for real problems.  

@trilinos/panzer 

@jmgate @rppawlo is this one of you?

## Expectations

Ideally the gather/scatter code would get cleaned up, there is 20k lines of it and it seems hard to maintain. but at least thread the blocked G/S code.

## Current Behavior

Saw a massive slowdown in my code when I went to blocked assy, 

## Definition of Done

 - [ ] thread the blocked codes
## Steps to Reproduce

1. Open code in edittor Panzer_ScatterResidual_BlockedTpetra.hpp Panzer_GatherSolution_BlockedTpetra_impl.hpp
1. Shake fist angrily at computer.
est log file:  
* When was the last time everything worked (date/time; SHA1s; etc.)?
* What did you do that made the bug rear its ugly head?
* Have you tried turning it off and on again?
-->

**Comment by rppawlo** [comment]
funny - was just going through the non-blocked version on the plane today. I can hit this up soon.

**Comment by rppawlo** [comment]
Have 4 classes due to specialization to convert:
- [ ] GatherSolution_BlockedTpetra (Residual)
- [ ] GatherSolution_BlockedTpetra (Jacobian)
- [ ] ScatterResidual_BlockedTpetra (Residual)
- [ ] ScatterResidual_BlockedTpetra (Jacobian)

Am punting on Tangent and Hessian for now. Looks like mini-em examples use this code, so we can use this to test changes in performance. As part of this refactor, I am hoping to remove the use of BlockedDOFManager and go directly to the sub-DOFManagers so that we can eventually deprecate the BlockedDOFManager. 

@eric-c-cyr @egphill - I haven't looked at how this might impact teko. Any comments?

**Comment by egphill** [comment]
@rppawlo Will this still result in a Thyra blocked linear op? If so, I don't think it will really effect Teko.

**Comment by rppawlo** [comment]
> Will this still result in a Thyra blocked linear op? If so, I don't think it will really effect Teko.

yes - the linear object factory will still produce the thyra blocked systems. Thanks Edward!
